### PR TITLE
Part 1: Add CSI go tests

### PR DESCRIPTION
**Pull request by dvonthenen**
**What this PR does / why we need it**:
We need more CSI tests! Tests added:
- Complete controller go test (Create, List, Publish, Unpublish, Delete)
- Test ListVolumes boundary conditions mainly around pagination
- Test ListVoumes order
- Test for Zones support using 2 Datacenters

Additional changes found while added go tests:
- updated govmomi to latest to fix a bug in device Key/UnitNumber in the older version
- Removed some klog.V(X) to always display since they are very useful debug statements to have all the time
- Changed some klog.V(X) values as some weren't needed as often and others were
- Found a bug in GetAllDatastoreClusters in which returns an error when no DatastoreClusters exist causing the entire ListVolumes to fail
- Changed ListVolumes to be on a 0-based index instead of 1-based. This is used for the NextToken when doing pagination of volumes. 1-based was too confusing.

**Which issue this PR fixes**:
NA

**Special notes for your reviewer**:
Tested CCM and CSI on:
k8s 1.13.2 cluster with 10 worker nodes in 3 zones
vSphere 6.7u1
Deployed a test pod which targeted a particular zone using a persistent volume
Used a multiple vCenters (2 to be exact) with multiple Datacenters (3 to be exact)
`make test` passes

**Release note**:
NA

### CODE REVIEW

File: pkg/csi/service/fcd/controller_test.go [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+	model.Service.TLS = tlsConfig
+	s := model.Service.NewServer()
+
+	// STS simulator
+	path, handler := sts.New(s.URL, vpx.Setting)
+	model.Service.ServeMux.Handle(path, handler)
+
+	// vAPI simulator
+	path, handler = vapi.New(s.URL, nil)
+	model.Service.ServeMux.Handle(path, handler)
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: Doesn't matter for this test, but you can use `model.Service.Handle` here.  Calls the same `ServeMux.Handle` underneath, but also connects the rest simulator to a soap simulator interface to support the [FCD tag APIs](https://pubs.vmware.com/vsphere-6-5/topic/com.vmware.wssdk.apiref.doc/vim.vslm.vcenter.VStorageObjectManager.html)

File: pkg/csi/service/fcd/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+	}
+	return configFromSim(multiDc)
+}
+
+type SearchIndex struct {
+	*simulator.SearchIndex
+	vm *simulator.VirtualMachine
+}
+
+func (s *SearchIndex) FindByDnsName(req *types.FindByDnsName) soap.HasFault {
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: With govmomi updated, that includes `FindByDnsName` support.  Though you still need to set `vm.Guest.HostName = vm.Name`, don't recall why we didn't do that by default.

File: pkg/csi/service/fcd/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+	// Get a simulator DS
+	myds := simulator.Map.Any("Datastore").(*simulator.Datastore)
+
+	err := connMgr.Connect(ctx, config.Global.VCenterIP)
+	if err != nil {
+		t.Errorf("Failed to Connect to vSphere: %s", err)
+	}
+
+	search := object.NewSearchIndex(connMgr.VsphereInstanceMap[config.Global.VCenterIP].Conn.Client)
+	si := simulator.Map.Get(search.Reference()).(*simulator.SearchIndex)
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: You can use `simulator.Map.SearchIndex()` here instead.

File: pkg/csi/service/fcd/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+	}
+
+	respCreate, err := c.CreateVolume(ctx, reqCreate)
+	if err != nil {
+		t.Fatalf("CreateVolume failed: %v", err)
+	}
+
+	volID := respCreate.Volume.VolumeId
+	volName := respCreate.Volume.VolumeContext[AttributeFirstClassDiskName]
+	log.Println("ID:", volID)
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: `t.Logf` instead so these only show up with `go test -v`.

File: pkg/csi/service/fcd/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+		}
+
+		volNameRet := respCreate.Volume.VolumeContext[AttributeFirstClassDiskName]
+		if !strings.EqualFold(volName, volNameRet) {
+			t.Errorf("[CREATE] Name of FCD does not match %s != %s", volName, volNameRet)
+		}
+	}
+
+	// get first 10
+	resp, err := c.ListVolumes(context.Background(), &csi.ListVolumesRequest{MaxEntries: 10})
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: You have a `ctx` var in scope that could be used in place of  `context.Background()` here and the ones below.

File: pkg/csi/service/fcd/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+	}
+
+	// Setup a multi-DC environment with zones!
+	// Setup DC0
+	dc0, err := vclib.GetDatacenter(ctx, vsi.Conn, "DC0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Attach a random tag to DC0
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: I think the "random" comment comes from the in-tree test, but isn't a random tag in these tests :)
https://github.com/kubernetes/kubernetes/blob/2f2945732b5821f9d51c6acdfb0e9fabe81bf3a4/pkg/cloudprovider/providers/vsphere/vsphere_test.go#L421-L422

File: pkg/csi/service/fcd/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,644 @@
...
+	if err = m.AttachTag(ctx, regionID, dc0.Reference()); err != nil {
+		t.Fatal(err)
+	}
+	if err = m.AttachTag(ctx, zoneIDwest, dc0.Reference()); err != nil {
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: No need to call `.Reference()` in any of the AttachTag calls, the param is a `mo.Reference` interface, which all the inventory types satisfy.

File: pkg/csi/service/fcd/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,644 @@
...
+	reqCreate.AccessibilityRequirements.Requisite = append(reqCreate.AccessibilityRequirements.Requisite, topology)
+
+	respCreate, err := c.CreateVolume(ctx, reqCreate)
+	if err != nil {
+		t.Fatalf("CreateVolume failed: %v", err)
+	}
+
+	volID := respCreate.Volume.VolumeId
+	volName := respCreate.Volume.VolumeContext[AttributeFirstClassDiskName]
+	log.Println("ID:", volID)
```
Comment (dougm) [inline pkg/csi/service/fcd/controller_test.go]: `t.Logf` here too.

File: pkg/csi/service/fcd/controller_test.go (hunk 4 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Yup, will change

File: pkg/csi/service/fcd/controller_test.go (hunk 5 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Copy and past error. Will update.

File: pkg/csi/service/fcd/controller_test.go (hunk 6 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: yup very true

File: pkg/csi/service/fcd/controller_test.go (hunk 8 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Yup

File: pkg/csi/service/fcd/controller_test.go (hunk 7 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Got it.

File: pkg/csi/service/fcd/controller_test.go [hunk 9]
```diff
@@ -0,0 +1,644 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (codenrhoden) [inline pkg/csi/service/fcd/controller_test.go]: year should be 2019 since this is a brand new file.

File: pkg/csi/service/fcd/controller_test.go (hunk 3 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Will update

File: pkg/csi/service/fcd/controller_test.go (hunk 9 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller_test.go]: Yup

**Comment by dvonthenen** [comment]
Updates based on feedback
